src/puppetlabs/puppetdb/mq_listener.clj

```diff
@@ -167,154 +151,98 @@
+  "Calls `update!` on the global and command specific metric for `k`"
+  [command version k v]
+  (update! (global-metric k) v)
+  (update! (cmd-metric command version k) v))
 
-     (mark! (global-metric :seen))
-     (time! (global-metric :processing-time)
-            (f msg))
-     (mark! (global-metric :processed))
+(defn call-with-command-metrics
+  "Invokes `f` including the related metrics updates"
+  [command version retries f]
+  (create-metrics-for-command! command version)
 
-     (catch fatal? {:keys [cause]}
-       (on-fatal msg cause)
-       (mark! (global-metric :fatal)))
+  (mark-both-metrics! command version :seen)
+  (update-both-metrics! command version :retry-counts retries)
 
-     (catch Throwable exception
-       (on-retry msg exception)
-       (mark! (global-metric :retried))))))
+  (mutils/multitime!
+   [(global-metric :processing-time)
+    (cmd-metric command version :processing-time)]
 
-(defn wrap-with-command-parser
-  "Wrap a message processor `f` such that all messages passed to `f` are
-  well-formed commands. If a message cannot be parsed, the `on-fatal` hook is
-  invoked, and `f` is ignored."
-  [f on-parse-error]
-  (fn [msg]
-    (let [parse-result (try-parse-command msg)]
-      (if (instance? Throwable parse-result)
-        (do
-          (mark! (global-metric :fatal))
-          (on-parse-error msg parse-result))
-        (f parse-result)))))
-
-(defn wrap-with-discard
-  "Wrap a message processor `f` such that incoming commands with a
-  retry count exceeding `max-retries` are discarded.
-
-  This assumes that all incoming messages are well-formed command
-  objects, such as those produced by the `wrap-with-command-parser`
-  middleware."
-  [f on-discard max-retries]
-  (fn [{:keys [command version annotations] :as msg}]
-    (let [retries (count (:attempts annotations))]
-      (create-metrics-for-command! command version)
-      (mark! (cmd-metric command version :seen))
-      (update! (global-metric :retry-counts) retries)
-      (update! (cmd-metric command version :retry-counts) retries)
-
-      (when (>= retries max-retries)
-        (mark! (global-metric :discarded))
-        (mark! (cmd-metric command version :discarded))
-        (on-discard msg))
-
-      (when (< retries max-retries)
-        (let [result (time! (cmd-metric command version :processing-time)
-                            (f msg))]
-          (mark! (cmd-metric command version :processed))
-          result)))))
-
-(defn wrap-with-thread-name
-  "Wrap a message processor `f` such that the calling thread's name is
-  set to `<prefix>-<thread-id>`. This is useful for situations where
-  thread names are ordinarily duplicated across multiple threads, or
-  if the default names aren't descriptive enough."
-  [f prefix]
+   (let [command-result (f)]
+     (mark-both-metrics! command version :processed)
+     command-result)))
+
+(defn parse-command
+  [msg]
+  (try+
+   (cmd/parse-command msg)
+   (catch AssertionError e
```

Comment:
Perhaps fine for now, but I wonder if we should revisit this and adjust parse-command to at least document this as part of its API, and perhaps consider having it return/throw something other than an AssertionError.